### ISSUE TITLE
Development docker setup fails due to issues with Webpack

### ISSUE THREAD

**Issue by saviorand** [outline/outline#2026]
Hi! I'm new here. Wanted to see if I can contribute with some additional embeds but ran into this issue.
Running Docker `make up` throws an error with "Module parse failed" in development.

**To Reproduce**
Follow exact instructions from [README](https://github.com/outline/outline#local-development) to reproduce the behavior.
1. Install required dependencies for development including Docker Desktop for Mac, clone github repo and run `make up`
2. See error:

```LOG from GenerateSW
<i> The service worker at service-worker.js will precache
<i>         5 URLs, totaling 306 kB.

ERROR in ./app/index.js 46:4
Module parse failed: Unexpected token (46:4)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| if (element) {
|   render(
>     <Provider {...stores}>
|       <Analytics>
|         <Theme>
 @ multi webpack-hot-middleware/client ./app/index main[1]
Child html-webpack-plugin for "index.html":
         Asset     Size  Chunks  Chunk Names
    index.html  1.4 MiB       0  
    Entrypoint undefined = index.html
    [../../../documents/github/outline/node_modules/html-webpack-plugin/lib/loader.js!./server/static/index.html] /Users/username/documents/github/outline/node_modules/html-webpack-plugin/lib/loader.js!./server/static/index.html 1.98 KiB {0} [built]
    [../../../documents/github/outline/node_modules/lodash/lodash.js] /Users/username/documents/github/outline/node_modules/lodash/lodash.js 530 KiB {0} [built]
    [./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {0} [built]
    [./node_modules/webpack/buildin/module.js] (webpack)/buildin/module.js 497 bytes {0} [built]
```

**Expected behavior**
Expected to install required dependencies and launch a dev server

**Outline (please complete the following information):**
- Install: self hosted
- Version (commit sha): 2a6dfdea5d5028ac46491caef79e98795b72cbd3

**Desktop (please complete the following information):**
 - OS: MacOS Big Sur 11.2
 - Docker Desktop version 3.3.0 (Apple Silicon)

Let me know if I'm doing something wrong here. Thanks!

**Comment by saviorand** [comment]
Tried several different webpack versions (^3.0.0, ^4.0.0), both fail due to different reasons. Not sure what's the exact issue here

**Comment by tommoor** [comment]
very strange, which version of node do you have locally? Perhaps try this branch:

https://github.com/outline/outline/pull/2024

**Comment by saviorand** [comment]
Hi @tommoor ! It's v. 12.22.1

**Comment by saviorand** [comment]
Been on a higher 14+ version earlier but downgraded after seeing the error ,didn't help

**Comment by saviorand** [comment]
@tommoor interesting, this #2024 branch led to the following output:

```
LOG from GenerateSW
<i> The service worker at service-worker.js will precache
<i>         5 URLs, totaling 900 kB.

WARNING in ./node_modules/ansi-html/index.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/ansi-html/index.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/client-overlay.js
* /Users/username/documents/github/outline/node_modules/ansi-html/index.js
    Used by 2 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/overlay/components/CompileErrorTrace.js
 @ ./node_modules/ansi-html/index.js
 @ (webpack)-hot-middleware/client-overlay.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in ./node_modules/ansi-regex/index.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/ansi-regex/index.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/strip-ansi/index.js
* /Users/username/documents/github/outline/node_modules/ansi-regex/index.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/strip-ansi/index.js
 @ ./node_modules/ansi-regex/index.js
 @ (webpack)-hot-middleware/node_modules/strip-ansi/index.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/client-overlay.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/client-overlay.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/client.js
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/client-overlay.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/client.js
 @ (webpack)-hot-middleware/client-overlay.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/client.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/client.js
    Used by 1 module(s), i. e.
    multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/client.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/sockets/WHMEventSource.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/node_modules/html-entities/lib/html4-entities.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/html4-entities.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/html4-entities.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
 @ (webpack)-hot-middleware/node_modules/html-entities/lib/html4-entities.js
 @ (webpack)-hot-middleware/node_modules/html-entities/lib/index.js
 @ (webpack)-hot-middleware/client-overlay.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/node_modules/html-entities/lib/html5-entities.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/html5-entities.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/html5-entities.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
 @ (webpack)-hot-middleware/node_modules/html-entities/lib/html5-entities.js
 @ (webpack)-hot-middleware/node_modules/html-entities/lib/index.js
 @ (webpack)-hot-middleware/client-overlay.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/node_modules/html-entities/lib/index.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/client-overlay.js
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/client-overlay.js
 @ (webpack)-hot-middleware/node_modules/html-entities/lib/index.js
 @ (webpack)-hot-middleware/client-overlay.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/node_modules/html-entities/lib/xml-entities.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/xml-entities.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/xml-entities.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/html-entities/lib/index.js
 @ (webpack)-hot-middleware/node_modules/html-entities/lib/xml-entities.js
 @ (webpack)-hot-middleware/node_modules/html-entities/lib/index.js
 @ (webpack)-hot-middleware/client-overlay.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/node_modules/strip-ansi/index.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/node_modules/strip-ansi/index.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/client.js
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/node_modules/strip-ansi/index.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/client.js
 @ (webpack)-hot-middleware/node_modules/strip-ansi/index.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

WARNING in (webpack)-hot-middleware/process-update.js
There are multiple modules with names that only differ in casing.
This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
Use equal casing. Compare these module identifiers:
* /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/process-update.js
    Used by 1 module(s), i. e.
    /Users/username/Documents/GitHub/outline/node_modules/webpack-hot-middleware/client.js
* /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/process-update.js
    Used by 1 module(s), i. e.
    /Users/username/documents/github/outline/node_modules/webpack-hot-middleware/client.js
 @ (webpack)-hot-middleware/process-update.js
 @ (webpack)-hot-middleware/client.js
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index

ERROR in ./app/index.js 49:4
Module parse failed: Unexpected token (49:4)
File was processed with these loaders:
 * ../../../documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/loader/index.js
You may need an additional loader to handle the result of these loaders.
| if (element) {
|   const App = () => (
>     <Provider {...stores}>
|       <Analytics>
|         <Theme>
 @ multi /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js webpack-hot-middleware/client /Users/username/documents/github/outline/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js ./app/index main[3]
```

**Comment by saviorand** [comment]
Guess there's some conflict of modules in place. Will try debugging this myself

**Comment by tommoor** [comment]
The main error suggests that it's not parsing JSX correctly, I don't know why that would be to be honest – it's baked into babelrc rules along with everything else.

**Comment by saviorand** [comment]
Couldn't figure it out 🤔 
My wild guess is it's connected to https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/259 somehow but couldn't confirm - versions of Node under 15.3.0 are reported not to play well with M1 Macs. Maybe I'll just try that Terraform script

**Comment by tommoor** [comment]
That dependency was only added in the branch I asked you to try though, it wasn't on the `main` branch two days ago – so it couldn't be that.

I can confirm that on an M1 you need Node 15+, I use an M1 myself.

**Comment by slurdge** [comment]
I have the exact same error on Windows 10.
```
ERROR in ./app/index.js 57:4
Module parse failed: Unexpected token (57:4)
File was processed with these loaders:
 * ./node_modules/@pmmmwh/react-refresh-webpack-plugin/loader/index.js
You may need an additional loader to handle the result of these loaders.
| if (element) {
|   const App = () => (
>     <Provider {...stores}>
|       <Analytics>
|         <Theme>
```

I did not use `make up`, instead I did every step by hand. Interestingly, I had also to copy the file `webpack.config.js` but the package.json has only `ln -sf \"$(pwd)/webpack.config.dev.js\" ./build` in its rules.

**Comment by tommoor** [comment]
> I did not use make up, instead I did every step by hand

Why? The scripts are provided to reduce issues like this

**Comment by slurdge** [comment]
> 
> 
> > I did not use make up, instead I did every step by hand
> 
> Why? The scripts are provided to reduce repetitive issues like this

I fully agree, I just don't have make available in my Windows terminal and didn't take the time to setup the WSL env. Actually I made myself a batch file which reproduces the steps afterwards.

**Comment by slurdge** [comment]
I did some more tests tonight.

1. I tried to build on WSL (Ubuntu) with latest yarn & node (16), I get the same error;
2. I tried to comment out the react refresh, but in this case, there is no loader for the file so it fails;
3. As per https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/136 I tried to force 'all' for '@babel/preset-flow'.
4. Tried to explicitly change NODE_ENV to 'development', while it changed messages from the server, it didn't impact front-end (I guess this is expected ?)

No changes I made, either from the Windows side or the Linux side, could make the latest version compile.
I'm at loss what to try next, would appreciate any pointer.

**Comment by tommoor** [comment]
I would review the development instructions again step by step. For development we don't run the node process inside of docker, is that what you're attempting to do?

**Comment by slurdge** [comment]
@tommoor Thank you for your time and help. I did follow the instructions but since some scripts steps don't work on Windows (namely cp & ln -s) I had to adapt.
In the end, I did a full Ubuntu VM and the instructions worked.

So I went back to Windows and discovered the difference: it is not possible to load the relative `  const config = require("../webpack.config.dev");` with a symlink on Windows. Therefore the root of the webpack config file is different, this is enough to make it trip over.
I fixed it by directly taking the original file `const config = require("../../webpack.config.dev");`, it now does work perfectly on my machine. I do not know why the symlink is done in the first place (instead of referencing the file), but on Windows this workaround is needed.
